File: AOC.Go/aoc2018/2018_04.go
package aoc2018

import (
	"regexp"
	"sort"
	"strconv"
)

func Solve2018_04_1(lines []string) int64 {
	infos := buildInfos(lines)

	var most *GuardInfo
	for _, v := range infos {
		if most == nil || v.totalMins > most.totalMins {
			most = v
		}
	}

	best := 0
	for i := 0; i < 60; i++ {
		if most.minuteCounts[i] > most.minuteCounts[best] {
			best = i
		}
	}

	return int64(most.id * (best))
}

func Solve2018_04_2(lines []string) int64 {
	infos := buildInfos(lines)

	var best *Part2Result

	for _, v := range infos {
		for i := 0; i < 60; i++ {
			if best == nil || v.minuteCounts[i] > best.count {
				result := Part2Result{
					id:     v.id,
					minute: i,
					count:  v.minuteCounts[i],
				}
				best = &result
			}
		}
	}

	return int64(best.id * best.minute)
}

func buildInfos(lines []string) map[int]*GuardInfo {
	guardRe := regexp.MustCompile(`Guard #(\d+)`)
	sleepRe := regexp.MustCompile(`:(\d\d)] falls`)
	wakeRe := regexp.MustCompile(`:(\d\d)] wakes`)

	sort.Strings(lines)

	guardInfos := make(map[int]*GuardInfo)

	var currInfo *GuardInfo
	var currSleep int

	for _, s := range lines {
		guardMatches := guardRe.FindStringSubmatch(s)
		sleepMatches := sleepRe.FindStringSubmatch(s)
		wakeMatches := wakeRe.FindStringSubmatch(s)

		if guardMatches != nil {
			found := false
			guardNum, _ := strconv.Atoi(guardMatches[1])
			currInfo, found = guardInfos[guardNum]
			if !found {
				newInfo := GuardInfo{
					id:        guardNum,
					totalMins: 0,
				}
				guardInfos[guardNum] = &newInfo
				currInfo = &newInfo
			}
		} else if sleepMatches != nil {
			currSleep, _ = strconv.Atoi(sleepMatches[1])
		} else if wakeMatches != nil {
			wakeTime, _ := strconv.Atoi(wakeMatches[1])
			mins := wakeTime - currSleep
			currInfo.totalMins += mins

			for i := currSleep; i < wakeTime; i++ {
				currInfo.minuteCounts[i]++
			}
		}
	}

	return guardInfos
}

type Part2Result struct {
	id     int
	minute int
	count  int
}

type GuardInfo struct {
	id           int
	totalMins    int
	minuteCounts [60]int
}

File: AOC.CSharp/2018/2018_04.cs
using System.Text.RegularExpressions;

namespace AOC.CSharp;

public static class AOC2018_04
{
    private static Regex GuardRe = new(@"Guard #(\d+)");
    private static Regex SleepRe = new(@":(\d\d)] falls");
    private static Regex WakeRe = new(@":(\d\d)] wakes");

    public static long Solve1(string[] lines)
    {
        var infos = BuildInfos(lines);
        var most = infos.MaxBy(kvp => kvp.Value.TotalMins).Value;
        int best = most.MinuteCounts.Select((count, idx) => (count, idx)).MaxBy(x => x.count).idx;

        return most.Id * best;
    }

    public static long Solve2(string[] lines)
    {
        var infos = BuildInfos(lines);

        Part2Result best = null;

        foreach (var info in infos.Values)
        {
            for (int i = 0; i < 60; i++)
            {
                if (best == null || info.MinuteCounts[i] > best.Count)
                {
                    best = new Part2Result(info.Id, i, info.MinuteCounts[i]);
                }
            }
        }

        return best.Id * best.Minute;
    }

    private static Dictionary<int, GuardInfo> BuildInfos(string[] lines)
    {
        Dictionary<int, GuardInfo> infos = new();

        GuardInfo currGuard = null;
        int currSleep = 0;
        foreach (var line in lines.OrderBy(x => x))
        {
            Match guardMatch = GuardRe.Match(line);
            Match sleepMatch = SleepRe.Match(line);
            Match wakeMatch = WakeRe.Match(line);

            if (guardMatch.Success)
            {
                int num = int.Parse(guardMatch.Groups[1].Value);
                currGuard = infos.TryGetValue(num, out GuardInfo g) ? g : null;
                if (currGuard == null)
                {
                    currGuard = new() { Id = num, TotalMins = 0, };
                    infos.Add(num, currGuard);
                }
            }
            else if (sleepMatch.Success)
            {
                currSleep = int.Parse(sleepMatch.Groups[1].Value);
            }
            else if (wakeMatch.Success)
            {
                int wake = int.Parse(wakeMatch.Groups[1].Value);
                int mins = wake - currSleep;
                currGuard.TotalMins += mins;

                for (int i = currSleep; i < wake; i++)
                {
                    currGuard.MinuteCounts[i]++;
                }
            }
        }

        return infos;
    }

    private record Part2Result(int Id, int Minute, int Count);

    private class GuardInfo
    {
        public int Id { get; set; }
        public int TotalMins { get; set; }
        public int[] MinuteCounts { get; } = new int[60];
    }
}
